src/v/utils/execution_monitor.h

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/vassert.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/gate.hh>
+#include <seastar/core/sstring.hh>
+#include <seastar/core/timer.hh>
+
+#include <cstdint>
+#include <optional>
+
+namespace ssx {
+
+template<typename Clock = seastar::lowres_clock>
+using execution_monitor_stall_callback
+  = std::function<void(basic_retry_chain_context<Clock>&)>;
+
+using execution_monitor_unexpected_shutdown_callback = std::function<void()>;
+
+/// The execution monitor is used to detect stalls in the execution
+/// of a process. It allows to register multiple callsites and checkpoint
+/// them periodically (when the callsite is reached).
+/// The set of callsites is fixed at the construction time. The callsites
+/// are constructed as a set of objects and passed to the c-tor.
+template<typename Clock = seastar::lowres_clock>
+class basic_execution_monitor {
+public:
+    using time_point = typename Clock::time_point;
+    using duration = typename Clock::duration;
+
+    /// Create a new execution monitor with a specified stall timeout.
+    basic_execution_monitor(
+      seastar::sstring name, duration stall_timeout) noexcept
+      : _name(std::move(name))
+      , _stall_timeout(stall_timeout) {}
+
+    basic_execution_monitor(const basic_execution_monitor&) = delete;
+    basic_execution_monitor(basic_execution_monitor&&) = delete;
+    basic_execution_monitor& operator=(const basic_execution_monitor&) = delete;
+    basic_execution_monitor& operator=(basic_execution_monitor&&) = delete;
+
+    struct stall_detector_config {
+        execution_monitor_stall_callback<Clock> cb;
+        std::vector<retry_chain_context*> contexts;
+    };
+
+    struct unexpected_shutdown_detector_config {
+        execution_monitor_unexpected_shutdown_callback cb;
```

Comment:
some documentation on these members would help to understand the code much easier

i.e. cb // callback that will be called if the monitored gate does not have any holders
gate // gate to monitor